### ISSUE TITLE
VoiceOver does not read empty line when arrowing down over it

### ISSUE THREAD

**Issue by adincebic** [microsoft/vscode#50584]
- VSCode Version: 1.23.1
- OS Version: macOS Siera 10.13.4

Steps to Reproduce:

There are no exact steps. The visual studio code is almost completely unusable with the screen reader (in this case VoiceOver). Navigating up and down through the file content is very difficult since it only reads the whole line and is not possible to go character by character or word by word. Also, very often the VoiceOver reads only line before but not the current one.
Intellisense sometimes works and sometimes does not.
In general, this is terribly done and yet Microsoft is constantly showing me the notification saying that visual studio code is now optimized for use with screen reader and even asks me to say whether I want to turn on that optimization which does not work at all. Even worse, the VS Code behaves the same with that feature on or off.

Does this issue occur when all extensions are disabled?: Yes/No
Yes

**Comment by alexandrudima** [comment]
@adincebic Can you please help me understand further what is working and what is not working?

For example, when I open xcode and use arrow up and arrow down, VoiceOver reads the entire line. The same happens when I open VS Code and use arrow up and arrow down, VoiceOver reads the entire line.

When I use arrow left and arrow right in VSCode, VoiceOver reads the character that the cursor moved over.

When I am using suggestions (i.e. Ctrl+Space), and I navigate through the suggestions using arrow up and arrow down, VoiceOver reads them out for me.

I cannot reproduce what you are reporting. Can you please distill a reproducible problem?

**Comment by adincebic** [comment]
@alexandrudima  Ok, let's try like this:
1. Turn voiceover on and launch VS Code
2. Try opening an existing folder using only the keyboard and the voiceover
3. If you managed to do so, try navigating to the sidebar and go through different files in that folder
4. Assuming that you managed to that too, now try interacting with the editor and write some code. When there is a few lines it is manageable, but notice when you are going up and down that the empty lines do not get read and that the voiceover speaks the line above or below.
5. While in the editor, watch how does the voiceover cursor moves when going through lines.
6. When you are at the top or the bottom of the file, try pressing up or down arrow again and you will hear the notification about the screen reader optimization.
7. Since we are talking about the notifications, there is no straight forward way of reading them except for stopping interaction with the editor and jumping through the interface to find them.

I think this is enough for now.

NOTE: Try to use VS Code without mouse, just like the real blind or visually impaired person.
NOTE: As you already mentioned Xcode, you can reference it as a almost perfectly accessible IDE for the blind. When I say almost, I mean that only the interface builder is not as accessible as it could be. From debugging through navigating the code base to autocomplete, everything is accessible.

I understand that this level of accessibility can be harder to achieve in VS Code since it is some kind of a web view application, but I am sure that these issues can be resolved.

**Comment by alexandrudima** [comment]
@adincebic It is very difficult to work with issues like the one you have created. Saying nothing works is really not helpful at all and not actionable (i.e. there will be no code change done so nothing will improve). Please create issues with clear steps, indicating what you do, what you expect to happen, and what VS Code does incorrectly in your opinion.

For example, I can open a folder by pressing Command+O and using the native picker to select a folder. From your sentence "Try opening an existing folder using only the keyboard and the voiceover" you are implying that something is not working. Can you please create a new issue where you, in detail, explain what is not working for you. That should be its own issue, with its own title and its own clear reproducible steps. Opening a folder is a completely separate area of VS Code from navigating with the arrows in the editor and deserves its own separate issue.

On point 3, I am able to press Command+Shift+E to bring focus to the file explorer and navigate through it using arrow keys. Alternatively, if focus is in a text editor and I wish to tab out of the editor in order to reach the Explorer viewlet, I can press Control+Shift+M to toggle that the editor traps Tab and then tab to the Explorer viewlet. Have you seen the documentation available at https://code.visualstudio.com/docs/editor/accessibility ? Again you are implying something is not working here. Please create a new issue, explaining what is not working for you. Perhaps the documentation is incomplete and has some gaps?

I can reproduce something odd when stepping over an empty line using arrow down. Here I will give an example of decent steps for an issue.

If the editor contains the following text:
```
1This is a file
2This is a file
3This is a file

4This is a file
5This is a file
6This is a file
7This is a file
```
And I position the cursor after the 1 in the first line:,
* pressing arrow down reads "2This is a file"
* pressing arrow down reads "3This is a file"
* pressing arrow down reads "1This is a file 2This is a file 3This is a file" instead of "newline", which is what XCode reads in this case.

I cannot reproduce anything weird regarding pressing arrow up or arrow down when at the top or at the bottom of a file.

So, for now, there is one issue -- the one I put in steps here -- that came out from your comment and that I can reproduce. 

Do you have any other issues that you would like to report (with clear steps please) ?

**Comment by adincebic** [comment]
@alexandrudima  you are right! I did not take a look at the documentation. For now, please try to fix the issue regarding reading the code line by line with the arrow keys.
I am going to make new issues with clear steps to reproduce.

Thank you for your feedback!

**Comment by amaniahmad** [comment]
Dear
Please can you help me 
how make website accessibility
step by step
thanks

**Comment by adincebic** [comment]
@amaniahmad  Making websites accessible consists of mainly few things:
# Properly structuring content. It means that you should use headings and other tags  wisely where they are needed (e.g. beginning of an article or section)
# Provide alternative descriptions for images with altText tag. The description is not visible but can be read by screen-readers. A common format should be like this: A kid with glasses sitting in-front of a computer.
# Provide "skip to content" links. These are use to jump to different sections of a web page. This allows the users to press on that link and get right to the content they want to read without sequentially going through the menus each time they visit your website.

There is a ton of resources available at:
https://www.w3.org/standards/webdesign/accessibility

Hope it helps!

**Comment by isidorn** [comment]
This issue feels like a duplicate of https://github.com/microsoft/vscode/issues/70033
If there are parts of this issue which are not covered by that one please create a new seperate issue so we can focus on those things seperatly. Thanks
